File: backend/app/schemas/demand.py
from datetime import date, datetime
from enum import Enum
from typing import Optional

from pydantic import BaseModel, ConfigDict, Field


class DemandStatus(str, Enum):
    open = "open"
    matched = "matched"
    fulfilled = "fulfilled"
    cancelled = "cancelled"


class DemandBase(BaseModel):
    buyer_id: int
    crop_name: str
    quantity: int = Field(gt=0)
    needed_by: Optional[date] = None


class DemandCreate(DemandBase):
    pass


class DemandStatusUpdate(BaseModel):
    status: DemandStatus


class Demand(DemandBase):
    model_config = ConfigDict(from_attributes=True)

    id: int
    status: DemandStatus = DemandStatus.open
    created_at: Optional[datetime] = None

File: src/lib/server/services/demand.ts
/**
 * Business logic for demand signals.
 *
 * Creating a demand publishes `buyer.request.created` so intelligence agents
 * can react to new purchasing intent.
 */

import { publish } from "../events";
import type { Demand, DemandStatus } from "../models";
import { buyers, demands, updateDemandStatus } from "../repositories";
import { utcnowIso } from "../time";
import { ConflictError, NotFoundError } from "./errors";

export interface DemandInput {
  buyer_id: number;
  crop_name: string;
  quantity: number;
  needed_by: string | null;
}

const ALLOWED_TRANSITIONS: Record<DemandStatus, DemandStatus[]> = {
  open: ["matched", "cancelled"],
  matched: ["fulfilled", "cancelled"],
  fulfilled: [],
  cancelled: [],
};

export function listDemands(options: {
  skip?: number;
  limit?: number;
  status?: DemandStatus | null;
}): Demand[] {
  return demands.list(options);
}

export function getDemand(demandId: number): Demand | null {
  return demands.get(demandId);
}

export async function createDemand(input: DemandInput): Promise<Demand> {
  if (buyers.get(input.buyer_id) === null) {
    throw new NotFoundError(`Buyer ${input.buyer_id} does not exist`);
  }

  const demand = demands.create({
    ...input,
    status: "open",
    created_at: utcnowIso(),
  });

  await publish("buyer.request.created", {
    demand_id: demand.id,
    buyer_id: demand.buyer_id,
    crop_name: demand.crop_name,
    quantity: demand.quantity,
  });
  return demand;
}

export function updateStatus(demand: Demand, newStatus: DemandStatus): Demand {
  if (!ALLOWED_TRANSITIONS[demand.status].includes(newStatus)) {
    throw new ConflictError(
      `Cannot move demand ${demand.id} from ${demand.status} to ${newStatus}`
    );
  }
  return updateDemandStatus(demand, newStatus);
}
